README.md
# flutter_overlay_apps

Android plugin for displaying flutter app over other apps

## Usage

Add dependency to pubspec.yaml file


### Android
You'll need to add the `SYSTEM_ALERT_WINDOW` permission and `OverlayService` to your Android Manifest.
```XML
    <uses-permission android:name="android.permission.SYSTEM_ALERT_WINDOW" />
    <application>
        ...
        <service
           android:name="com.phan_tech.flutter_overlay_apps.OverlayService"
           android:exported="false" />
    </application>
```

### Entry point
Inside `main.dart` create an entry point for your Overlay widget;
NOTE: `MaterialApp` is required
```dart
// overlay entry point
@pragma("vm:entry-point")
void showOverlay() {
  runApp(const MaterialApp(
    debugShowCheckedModeBanner: false,
    home: Material(child: Text("My overlay"))
  ));
}
```


### Methods
To open an overlay, call `FlutterOverlayApps.showOverlay()`. 
Default `height` & `width` is fill screen

```dart
FlutterOverlayApps.showOverlay(height: 300, width: 400, alignment: OverlayAlignment.center);
```

To close the overlay widget call 
```dart
FlutterOverlayApps.closeOverlay();
```
To send data to and from Overlay widget, call 
```dart
FlutterOverlayApps.sendDataToAndFromOverlay(<data>);
```
For listening to broadcasted data, stream the messages by calling 
```dart
FlutterOverlayApps.overlayListener().listen((data) {
    print(data);
});
```

### Code Example
```dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter_overlay_apps/flutter_overlay_apps.dart';

void main() {
  runApp(const MyApp());
}

// overlay entry point
@pragma("vm:entry-point")
void showOverlay() {
  runApp(const MaterialApp(
    debugShowCheckedModeBanner: false,
    home: MyOverlaContent()
  ));
}


class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: ElevatedButton(
            onPressed: () async {

              // Open overlay
              await FlutterOverlayApps.showOverlay(height: 300, width: 400, alignment: OverlayAlignment.center);

              // send data to ovelay
              await Future.delayed(const Duration(seconds: 2));
              FlutterOverlayApps.sendDataToAndFromOverlay("Hello from main app");
            }, 
            child: const Text("showOverlay")
          ),
        ),
      ),
    );
  }
}



class MyOverlaContent extends StatefulWidget {
  const MyOverlaContent({ Key? key }) : super(key: key);

  @override
  State<MyOverlaContent> createState() => _MyOverlaContentState();
}

class _MyOverlaContentState extends State<MyOverlaContent> {
  String _dataFromApp = "Hey send data";

  @override
  void initState() {
    super.initState();

    // lisent for any data from the main app
    FlutterOverlayApps.overlayListener().listen((event) {
      setState(() {
        _dataFromApp = event.toString();
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return Material(
      child: InkWell(
        onTap: (){
          // close overlay
          FlutterOverlayApps.closeOverlay();
        },
        child: Card(
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(16),
          ),
          child: Center(child: Text(_dataFromApp, style: const TextStyle(color: Colors.red),)),
        ),
      ),
    );
  }
}
```

## Android back button  
The back navigation button on android will close the overlay app. To prevent this, set `closeOnBackButton` to `false`. Example;

```dart
await FlutterOverlayApps.showOverlay(
  height: 300,
  closeOnBackButton: false,
  alignment: OverlayAlignment.center);
```

When `closeOnBackButton` is set to `false`, an action will be sent via the stream controller. You can get the action like the example below;

```dart
var overlayStreamController = FlutterOverlayApps.overlayListener();

overlayStreamController.stream.listen((event) {
  if(event['method'] == 'backButton'){
    // handle back button 
    if(mounted && Navigator.of(context).canPop()){
      Navigator.of(context).pop();
    }
  }
})
```

Support the plugin <a href="https://www.buymeacoffee.com/EddieGenius" target="_blank"><img src="https://i.imgur.com/aV6DDA7.png" alt="Buy Me A Coffee" style="height: 41px !important;width: 174px !important; box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;-webkit-box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;" > </a>


CHANGELOG.md
## 1.2.0

* Upgraded gradle
* Added Back Button navigation action

## 1.1.2

* exposed overlay StreamController

## 1.0.2

* enabled keyboard input

## 1.0.1

* Update app from overlay fix

## 1.0.0

* Ininial release


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_overlay_apps_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/flutter_overlay_apps_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_overlay_apps/flutter_overlay_apps.dart';

void main() {
  const MethodChannel channel =
      MethodChannel('com.phan_tech./flutter_overlay_apps');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return true;
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await FlutterOverlayApps.showOverlay(), true);
  });
}


lib/flutter_overlay_apps.dart
import 'dart:async';

import 'package:flutter/services.dart';

const int overlaySizeFill = -1;
const String _mainAppMethodChannel = "com.phan_tech./flutter_overlay_apps";
const String _overlayAppMethodChannel =
    "com.phan_tech/flutter_overlay_apps/overlay";
const String _overlayAppMessageChannel =
    "com.phan_tech/flutter_overlay_apps/overlay/messenger";

class FlutterOverlayApps {
  static const MethodChannel _channel = MethodChannel(_mainAppMethodChannel);

  // overlay methodChanel
  static const MethodChannel _overlayChannel =
      MethodChannel(_overlayAppMethodChannel);
  //Overlay BasicMessageChannel
  static const BasicMessageChannel _overlayMessageChannel =
      BasicMessageChannel(_overlayAppMessageChannel, JSONMessageCodec());

  /// Open overLay content
  /// Takes optional;
  ///   - int [height] default is [overlaySizeFill]
  ///   - int [width] default is [overlaySizeFill]
  ///   - OverlayAlignment [width] default is [alignment] [OverlayAlignment.center]
  ///   - bool [closeOnBackButton] default is `true`
  static Future<bool?> showOverlay(
      {int height = overlaySizeFill,
      int width = overlaySizeFill,
      bool closeOnBackButton = true,
      OverlayAlignment alignment = OverlayAlignment.center}) async {
    final bool? _res = await _channel.invokeMethod('showOverlay', {
      "height": height,
      "width": width,
      "alignment": alignment.name,
      "closeOnBackButton": closeOnBackButton
    });
    return _res;
  }

  /// Closes overlau if open
  static Future<bool?> closeOverlay() async {
    final bool? _res = await _overlayChannel.invokeMethod('close');
    return _res;
  }

  /// broadcast data to and from overlay app
  /// the supported data type are;
  ///   - [int], [double], [bool], [String], null
  ///   - [List] of supported types
  ///   - [Map] of supported types
  static Future sendDataToAndFromOverlay(dynamic data) async {
    return await _overlayMessageChannel.send(data);
  }

  /// Streams message shared between overlay and main app
  static final StreamController _controller = StreamController();
  static StreamController overlayListener() {
    _overlayMessageChannel.setMessageHandler((message) async {
      _controller.add(message);
      return message;
    });
    return _controller;
  }

  /// dispose overlay controller
  static void disposeOverlayListener() {
    _controller.close();
  }
}

/// Overlay alignment on screen
enum OverlayAlignment {
  topLeft,
  topCenter,
  topRight,
  centerLeft,
  center,
  centerRight,
  bottomLeft,
  bottomCenter,
  bottomRight
}


android/src/main/kotlin/com/phan_tech/flutter_overlay_apps/WindowSetup.kt
package com.phan_tech.flutter_overlay_apps

import android.annotation.SuppressLint
import android.view.Gravity
import android.view.WindowManager
import io.flutter.plugin.common.BasicMessageChannel

object WindowSetup {
    var height: Int = -1
    var width: Int = WindowManager.LayoutParams.MATCH_PARENT
    var gravity: Int = Gravity.CENTER
    var messenger : BasicMessageChannel<Any?>? = null
    var closeOnBackButton : Boolean = true;

    @SuppressLint("RtlHardcoded")
    fun setGravityFromAlignment(alignment: String){
        when {
            alignment.lowercase() == "topLeft".lowercase() -> {
                gravity = Gravity.TOP or Gravity.LEFT
            }
            alignment.lowercase() == "topCenter".lowercase() -> {
                gravity = Gravity.TOP
            }
            alignment.lowercase() == "topRight".lowercase() -> {
                gravity = Gravity.TOP or Gravity.RIGHT
            }


            alignment.lowercase() == "centerLeft".lowercase() -> {
                gravity = Gravity.CENTER or Gravity.LEFT
            }
            alignment.lowercase() == "center".lowercase() -> {
                gravity = Gravity.CENTER
            }
            alignment.lowercase() == "centerRight".lowercase() -> {
                gravity = Gravity.CENTER or Gravity.RIGHT
            }


            alignment.lowercase() == "bottomLeft".lowercase() -> {
                gravity = Gravity.BOTTOM or Gravity.LEFT
            }
            alignment.lowercase() == "bottomCenter".lowercase() -> {
                gravity = Gravity.BOTTOM
            }
            alignment.lowercase() == "bottomRight".lowercase() -> {
                gravity = Gravity.BOTTOM or Gravity.RIGHT
            }


        }
    }
}

android/src/main/kotlin/com/phan_tech/flutter_overlay_apps/OverlayService.kt
package com.phan_tech.flutter_overlay_apps

import android.app.Service
import android.content.Intent
import android.graphics.PixelFormat
import android.os.Build
import android.os.IBinder
import android.view.KeyEvent
import android.view.WindowManager
import android.widget.FrameLayout
import androidx.annotation.RequiresApi
import io.flutter.embedding.android.FlutterView
import io.flutter.embedding.engine.FlutterEngineCache
import io.flutter.plugin.common.BasicMessageChannel
import io.flutter.plugin.common.JSONMessageCodec
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject


class OverlayService : Service() {
    private var windowManager: WindowManager? = null
    private lateinit var flutterView: FlutterView
    private val flutterChannel = MethodChannel(FlutterEngineCache.getInstance().get("my_engine_id")!!.dartExecutor, overlayAppMethodChannel)
    private val overlayMessageChannel = BasicMessageChannel(FlutterEngineCache.getInstance().get("my_engine_id")!!.dartExecutor, overlayAppMessageChannel, JSONMessageCodec.INSTANCE)

    override fun onBind(intent: Intent?): IBinder? {
        // Not used
        return null
    }

    @RequiresApi(Build.VERSION_CODES.O)
    override fun onCreate() {
        super.onCreate()

        val engine = FlutterEngineCache.getInstance().get("my_engine_id")!!
        engine.lifecycleChannel.appIsResumed()

        flutterView = object: FlutterView(applicationContext){
            override fun dispatchKeyEvent(event: KeyEvent): Boolean {
                return if (event.keyCode == KeyEvent.KEYCODE_BACK) {
                    // handle the back button code;
                    if(WindowSetup.closeOnBackButton){
                        stopService(Intent(baseContext, OverlayService().javaClass))
                        windowManager?.removeView(flutterView)
                    }else{
                        // send message
                        overlayMessageChannel.send(JSONObject("{\"method\": \"backButton\"}"))// {"method" "backButton"}

                    }

                    true
                } else super.dispatchKeyEvent(event)
            }

        }

        flutterView.attachToFlutterEngine(FlutterEngineCache.getInstance().get("my_engine_id")!!)
        flutterView.fitsSystemWindows = true

        flutterChannel.setMethodCallHandler{ methodCall: MethodCall, result: MethodChannel.Result ->
            if(methodCall.method == "close"){
                val closed = stopService(Intent(baseContext, OverlayService().javaClass))
                result.success(closed)
            }
        }
        overlayMessageChannel.setMessageHandler(MyHandler())


        windowManager = getSystemService(WINDOW_SERVICE) as WindowManager?

        val params = WindowManager.LayoutParams(
            WindowSetup.width,
            WindowSetup.height,
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
            else WindowManager.LayoutParams.TYPE_SYSTEM_OVERLAY,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
            PixelFormat.TRANSLUCENT
        )
        params.flags = params.flags and WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE.inv()
        params.gravity = WindowSetup.gravity
        windowManager!!.addView(flutterView, params)
    }


    override fun onDestroy() {
        super.onDestroy()
        windowManager!!.removeView(flutterView)
    }
}

class MyHandler: BasicMessageChannel.MessageHandler<Any?>{
    override fun onMessage(message: Any?, reply: BasicMessageChannel.Reply<Any?>) {
        WindowSetup.messenger!!.send(message)
    }

}

android/src/main/kotlin/com/phan_tech/flutter_overlay_apps/FlutterOverlayAppsPlugin.kt
package com.phan_tech.flutter_overlay_apps

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.Settings
import androidx.annotation.NonNull
import androidx.annotation.RequiresApi
import io.flutter.FlutterInjector
import io.flutter.embedding.engine.FlutterEngineCache
import io.flutter.embedding.engine.FlutterEngineGroup
import io.flutter.embedding.engine.dart.DartExecutor

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.*
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

const val mainAppMethodChannel: String = "com.phan_tech./flutter_overlay_apps"
const val overlayAppMethodChannel: String = "com.phan_tech/flutter_overlay_apps/overlay"
const val overlayAppMessageChannel: String = "com.phan_tech/flutter_overlay_apps/overlay/messenger"

/** FlutterOverlayAppsPlugin */
class FlutterOverlayAppsPlugin: FlutterPlugin, MethodCallHandler, ActivityAware, BasicMessageChannel.MessageHandler<Any?> {
  private lateinit var channel : MethodChannel
  private lateinit var messenger : BasicMessageChannel<Any?>
  private lateinit var context: Context
  private lateinit var activity: Activity

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    WindowSetup.messenger = messenger
    WindowSetup.messenger!!.setMessageHandler(this)
    activity = binding.activity
    val enn = FlutterEngineGroup(context)
    val dEntry = DartExecutor.DartEntrypoint(
      FlutterInjector.instance().flutterLoader().findAppBundlePath(),
      "showOverlay")

    val engine = enn.createAndRunEngine(context, dEntry)

    FlutterEngineCache.getInstance().put("my_engine_id", engine)

  }

  override fun onDetachedFromActivityForConfigChanges() {
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
  }

  override fun onDetachedFromActivity() {
  }

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, mainAppMethodChannel)
    channel.setMethodCallHandler(this)

    messenger = BasicMessageChannel(flutterPluginBinding.binaryMessenger, overlayAppMessageChannel, JSONMessageCodec.INSTANCE)
    messenger.setMessageHandler(this)
    this.context = flutterPluginBinding.applicationContext

  }



  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "showOverlay" -> {

        // get permissions
        if(Build.VERSION.SDK_INT < Build.VERSION_CODES.M){
          result.error("1","SDK version is lower than 23", "Requires android sdk 23 and above")
        }
        else if(!checkPermissions()){
          requestPermissions()
        }else{
          val h = call.argument<Int>("height")
          val w = call.argument<Int>("width")
          val alignment = call.argument<String>("alignment")
          val closeOnBackButton = call.argument<Boolean>("closeOnBackButton")

          WindowSetup.width = w ?: -1
          WindowSetup.height = h ?: -1
          WindowSetup.closeOnBackButton = closeOnBackButton ?: true
          WindowSetup.setGravityFromAlignment(alignment ?: "center")
          activity.startService(Intent(context, OverlayService().javaClass))
          result.success(true)
        }


      }
//      "closeOverlay" -> {
//        val overlayChannel = MethodChannel(FlutterEngineCache.getInstance().get("my_engine_id")!!.dartExecutor, "com.phan_tech/overlay")
//        overlayChannel.invokeMethod("close", null)
//      }
      else -> {
        result.notImplemented()
      }
    }
  }

  override fun onMessage(message: Any?, reply: BasicMessageChannel.Reply<Any?>) {
    val overlayMessageChannel = BasicMessageChannel(FlutterEngineCache.getInstance().get("my_engine_id")!!.dartExecutor, overlayAppMessageChannel, JSONMessageCodec.INSTANCE)
    overlayMessageChannel.send(message, reply)
  }

  @RequiresApi(Build.VERSION_CODES.M)
  private fun checkPermissions(): Boolean{
    return Settings.canDrawOverlays(context)

  }

  @RequiresApi(Build.VERSION_CODES.M)
  private fun requestPermissions(){
    activity.startActivity(
      Intent(
        Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
        Uri.parse("package:${activity.packageName}"))
    )
  }



}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter_overlay_apps/flutter_overlay_apps.dart';

void main() {
  runApp(const MyApp());
}

// overlay entry point
@pragma("vm:entry-point")
void showOverlay() {
  runApp(const MaterialApp(
      debugShowCheckedModeBanner: false, home: MyOverlaContent()));
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: ElevatedButton(
              onPressed: () async {
                // Open overlay
                await FlutterOverlayApps.showOverlay(
                    height: 300,
                    width: 400,
                    alignment: OverlayAlignment.center);

                // send data to ovelay
                await Future.delayed(const Duration(seconds: 2));
                FlutterOverlayApps.sendDataToAndFromOverlay(
                    "Hello from main app");
              },
              child: const Text("showOverlay")),
        ),
      ),
    );
  }
}

class MyOverlaContent extends StatefulWidget {
  const MyOverlaContent({Key? key}) : super(key: key);

  @override
  State<MyOverlaContent> createState() => _MyOverlaContentState();
}

class _MyOverlaContentState extends State<MyOverlaContent> {
  String _dataFromApp = "Hey send data";
  late final StreamController overlayStream;

  @override
  void initState() {
    super.initState();

    // lisent for any data from the main app
    overlayStream = FlutterOverlayApps.overlayListener();
    overlayStream.stream.listen((event) {
      setState(() {
        _dataFromApp = event.toString();
      });
    });
  }

  @override
  void dispose() {
    overlayStream.close();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Material(
      child: InkWell(
        onTap: () {
          // close overlay
          FlutterOverlayApps.closeOverlay();
        },
        child: Card(
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(16),
          ),
          child: Center(
              child: Text(
            _dataFromApp,
            style: const TextStyle(color: Colors.red),
          )),
        ),
      ),
    );
  }
}
